pypck/module.py

```diff
@@ -305,6 +305,122 @@ def oem_text(self) -> List[str]:
         # return ''.join([block for block in self._oem_text if block])
 
 
+class GroupMembershipRequestHandler:
+    """Request handler to request static and dynamic group membership of a module."""
+
+    def __init__(
+        self,
+        addr_conn: "ModuleConnection",
+        num_tries: int = 3,
+        timeout_msec: int = 1500,
+    ):
+        """Initialize class instance."""
+        self.addr_conn = addr_conn
+
+        self._static_groups: List[LcnAddr] = []
+        self._dynamic_groups: List[LcnAddr] = []
+
+        self.static_groups_trh = TimeoutRetryHandler(num_tries, timeout_msec)
+        self.static_groups_trh.set_timeout_callback(self.timeout_static_groups)
+        self.dynamic_groups_trh = TimeoutRetryHandler(num_tries, timeout_msec)
+        self.dynamic_groups_trh.set_timeout_callback(self.timeout_dynamic_groups)
+
+        # callback
+        addr_conn.register_for_inputs(self.process_input)
+
+        # events
+        self.static_groups_known = asyncio.Event()
+        self.dynamic_groups_known = asyncio.Event()
+
+    def process_input(self, inp: inputs.Input) -> None:
+        """Create a task to process the input object concurrently."""
+        asyncio.create_task(self.async_process_input(inp))
+
+    async def async_process_input(self, inp: inputs.Input) -> None:
+        """Process incoming input object.
+
+        Method to handle incoming commands for this specific request handler.
+        """
+        if isinstance(inp, inputs.ModStatusGroups):
+            if inp.dynamic:
+                self._dynamic_groups = inp.groups
+                self.dynamic_groups_known.set()
```

Comment:
I rechecked it. To avoid an assertion error we have to check if the event is already set. ALso we have to cancel the corresponding TimeoutRetryHandler:
```python3
if not self.dynamic_groups_known.is_set():
    self.dynamic_groups_known.set()
await self.cancel_dynamic_groups()
```

Reply:
Hm `self.dynamic_groups_known.set()` should do nothing id the Event is already set.  At least the docs page does not indicate that it is not allowed to call `set()` if `is_set()` is already.

The cancel definitely should be called. So I would assume the following should work:

```
self.dynamic_groups_known.set()
await self.cancel_dynamic_groups()
```

I tried to repro the assert but couldn't trigger it. Otherwise I'd add a test for it.

Reply:
Yes, you are right. Prior to Events I used Futures. Future.set_result() raises an exception if already set. Events don't.